File: packages/core/src/stages/utils/srt.rs
//! SRT 写出 (镜像 TS `packages/core/utils/srt.ts` 的 `writeSrt` + `validateSrtContent`)。
//!
//! 仅实现「基础分支」: 每段输出一条 SRT (TS 的 `splitSubtitle` 多 fragment 切分逻辑未移植,
//! 因当前 Rust 流程的 segments 已是合适粒度, 与一致行为足够)。

use std::path::Path;

/// 单条 SRT 输入 (调用方负责从各 stage 结果映射到该结构)。
#[derive(Debug, Clone)]
pub struct SrtSeg {
    pub start_ms: u64,
    pub end_ms: u64,
    /// 翻译后文本 (优先); use_source=true 时改用 source
    pub dst: String,
    /// 原文 (use_source 时使用)
    pub text: String,
    /// 实际开始/结束时间 (Timing 携带, 覆盖 start_ms/end_ms)
    pub actual_start: Option<u64>,
    pub actual_end: Option<u64>,
}

/// 毫秒 → SRT 时间 `HH:MM:SS,mmm` (镜像 TS `srtTime`)。
fn srt_time(ms: u64) -> String {
    let total_secs = ms / 1000;
    let hh = total_secs / 3600;
    let mm = (total_secs % 3600) / 60;
    let ss = total_secs % 60;
    let mmm = ms % 1000;
    format!("{hh:02}:{mm:02}:{ss:02},{mmm:03}")
}

/// 写 SRT 文件 (镜像 TS `writeSrt`)。
///
/// - `use_source=true` 用 `text` (原文), 否则用 `dst` (翻译)
/// - 写盘前做 [`validate_srt_content`] 预检, 避免 ffmpeg 报含糊的 "Unable to open"
pub fn write_srt(segments: &[SrtSeg], output: &Path, use_source: bool) -> anyhow::Result<()> {
    let mut lines: Vec<String> = Vec::new();
    let mut idx: u32 = 1;
    for item in segments {
        let start = item.actual_start.unwrap_or(item.start_ms);
        let end = item.actual_end.unwrap_or(item.end_ms);
        let text = if use_source {
            item.text.trim().to_string()
        } else {
            let t = item.dst.trim();
            if t.is_empty() {
                item.text.trim().to_string()
            } else {
                t.to_string()
            }
        };
        if text.is_empty() {
            continue;
        }
        lines.push(idx.to_string());
        lines.push(format!("{} --> {}", srt_time(start), srt_time(end)));
        lines.push(text);
        lines.push(String::new());
        idx += 1;
    }
    let content = lines.join("\n");
    validate_srt_content(&content, output)?;
    std::fs::write(output, content)
        .map_err(|e| anyhow::anyhow!("写入 SRT {} 失败: {e}", output.display()))?;
    Ok(())
}

/// SRT 预检 (镜像 TS `validateSrtContent`): 在交给 ffmpeg 前拦截内容问题。
pub fn validate_srt_content(content: &str, file_path: &Path) -> anyhow::Result<()> {
    let raw_lines: Vec<&str> = content.split('\n').collect();
    let mut blocks: Vec<Vec<&str>> = Vec::new();
    let mut current: Vec<&str> = Vec::new();
    for line in raw_lines {
        if line.trim().is_empty() {
            if !current.is_empty() {
                blocks.push(std::mem::take(&mut current));
            }
        } else {
            current.push(line);
        }
    }
    if !current.is_empty() {
        blocks.push(current);
    }

    if blocks.is_empty() {
        return Err(anyhow::anyhow!(
            "SRT 预检失败 ({}): 文件为空，没有任何字幕块",
            file_path.display()
        ));
    }

    for (i, block) in blocks.iter().enumerate() {
        let block_no = i + 1;
        if block.len() < 2 {
            return Err(anyhow::anyhow!(
                "SRT 预检失败 ({}): 第 {} 块结构非法，期望 \"序号 / 时间轴 / 文本\"，实际 {} 行",
                file_path.display(),
                block_no,
                block.len()
            ));
        }
        let parts: Vec<&str> = block[1].split(" --> ").collect();
        if parts.len() != 2 {
            return Err(anyhow::anyhow!(
                "SRT 预检失败 ({}): 第 {} 块时间轴分隔符非法",
                file_path.display(),
                block_no
            ));
        }
        if !is_srt_time(parts[0]) || !is_srt_time(parts[1]) {
            return Err(anyhow::anyhow!(
                "SRT 预检失败 ({}): 第 {} 块时间轴格式非法: \"{}\" (应为 HH:MM:SS,mmm --> HH:MM:SS,mmm)",
                file_path.display(),
                block_no,
                block[1]
            ));
        }
        let s = parse_srt_time(parts[0]);
        let e = parse_srt_time(parts[1]);
        if !(e > s) {
            return Err(anyhow::anyhow!(
                "SRT 预检失败 ({}): 第 {} 块时间轴非法，结束时间必须晚于开始时间 ({})",
                file_path.display(),
                block_no,
                block[1]
            ));
        }
        if block[2].contains('\0') {
            return Err(anyhow::anyhow!(
                "SRT 预检失败 ({}): 第 {} 块文本含 NUL 字符",
                file_path.display(),
                block_no
            ));
        }
    }
    Ok(())
}

/// 校验 `HH:MM:SS,mmm` 格式 (无正则依赖)。
fn is_srt_time(t: &str) -> bool {
    let parts: Vec<&str> = t.split(':').collect();
    if parts.len() != 3 {
        return false;
    }
    let hh = parts[0].parse::<u32>().ok();
    let mm = parts[1].parse::<u32>().ok();
    let (ss, ms) = {
        let rest: Vec<&str> = parts[2].split(',').collect();
        if rest.len() != 2 {
            return false;
        }
        (rest[0].parse::<u32>().ok(), rest[1].parse::<u32>().ok())
    };
    hh.is_some() && mm.is_some() && ss.is_some() && ms.is_some() && ms.unwrap() < 1000
}

fn parse_srt_time(t: &str) -> u64 {
    let parts: Vec<&str> = t.split(':').collect();
    if parts.len() != 3 {
        return 0;
    }
    let h: u64 = parts[0].parse().unwrap_or(0);
    let m: u64 = parts[1].parse().unwrap_or(0);
    let rest: Vec<&str> = parts[2].split(',').collect();
    let s: u64 = rest.first().and_then(|x| x.parse().ok()).unwrap_or(0);
    let ms: u64 = rest.get(1).and_then(|x| x.parse().ok()).unwrap_or(0);
    (h * 3600 + m * 60 + s) * 1000 + ms
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    fn seg(start: u64, end: u64, dst: &str, text: &str) -> SrtSeg {
        SrtSeg {
            start_ms: start,
            end_ms: end,
            dst: dst.to_string(),
            text: text.to_string(),
            actual_start: None,
            actual_end: None,
        }
    }

    #[test]
    fn write_srt_roundtrip() {
        let dir = std::env::temp_dir()
            .join(format!("ld_srt_{}", std::process::id()))
            .to_string_lossy()
            .to_string();
        std::fs::create_dir_all(&dir).unwrap();
        let p = std::path::Path::new(&dir).join("out.srt");
        let segs = vec![
            seg(0, 1000, "你好", "hello"),
            seg(1000, 2500, "世界", "world"),
        ];
        write_srt(&segs, &p, false).unwrap();
        let content = std::fs::read_to_string(&p).unwrap();
        assert!(content.contains("你好"));
        assert!(content.contains("世界"));
        assert!(content.contains("00:00:00,000 --> 00:00:01,000"));
        assert!(content.contains("00:00:01,000 --> 00:00:02,500"));
        // 块号自增
        assert!(content.starts_with("1\n"));
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn write_srt_use_source() {
        let dir = std::env::temp_dir()
            .join(format!("ld_srt_src_{}", std::process::id()))
            .to_string_lossy()
            .to_string();
        std::fs::create_dir_all(&dir).unwrap();
        let p = std::path::Path::new(&dir).join("out.srt");
        // dst 空 → 回退 text
        let segs = vec![seg(0, 1000, "", "原文")];
        write_srt(&segs, &p, false).unwrap();
        let content = std::fs::read_to_string(&p).unwrap();
        assert!(content.contains("原文"));
        // use_source=true → 用 text
        let segs2 = vec![seg(0, 1000, "译文", "原文")];
        write_srt(&segs2, &p, true).unwrap();
        let content2 = std::fs::read_to_string(&p).unwrap();
        assert!(content2.contains("原文"));
        assert!(!content2.contains("译文"));
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn validate_rejects_bad_time_order() {
        let dir = std::env::temp_dir()
            .join(format!("ld_srt_bad_{}", std::process::id()))
            .to_string_lossy()
            .to_string();
        std::fs::create_dir_all(&dir).unwrap();
        let p = std::path::Path::new(&dir).join("bad.srt");
        // 结束早于开始
        let content = "1\n00:00:02,000 --> 00:00:01,000\n文本\n".to_string();
        let f = std::fs::File::create(&p).unwrap();
        let mut w = std::io::BufWriter::new(f);
        w.write_all(content.as_bytes()).unwrap();
        drop(w);
        let err = validate_srt_content(&content, &p).unwrap_err();
        assert!(err.to_string().contains("结束时间必须晚于开始时间"));
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn srt_time_format() {
        assert_eq!(srt_time(0), "00:00:00,000");
        assert_eq!(srt_time(1000), "00:00:01,000");
        assert_eq!(srt_time(61_250), "00:01:01,250");
        assert_eq!(srt_time(3_661_009), "01:01:01,009");
    }
}

File: packages/core/utils/srt.ts
import { TaskCtx } from "../context/context";
import { srtTime } from "./utils";
import { writeFile } from "./fileOps";
import { Timing } from "../stages/mix_audio/types";
import { TranslateSegment } from "../stages/05_translate/out";

/**
 *  按标点切分长句（,, ,, 。, ? 等），但保护配对符号（《》、「」 里的内容不被切开）
 */
function splitProtected(text: string): string[] {
  const PUNCTUATION = new Set(["，", ",", "；", ";", "：", ":", "。", "?", "？", "!", "！", "、"]);
  const PROTECTED_PAIRS: Record<string, string> = {
    "《": "》",
    "（": "）",
    "【": "】",
    "「": "」",
    "『": "』",
  };
  const segs: string[] = [];
  let buf: string[] = [],
    inside: string | null = null;
  for (const ch of text) {
    if (!inside && ch in PROTECTED_PAIRS) {
      inside = PROTECTED_PAIRS[ch];
      buf.push(ch);
      continue;
    }
    if (inside && ch === inside) {
      inside = null;
      buf.push(ch);
      continue;
    }
    if (!inside && PUNCTUATION.has(ch)) {
      const s = buf.join("").trim();
      if (s) segs.push(s);
      buf = [];
      continue;
    }
    buf.push(ch);
  }
  const tail = buf.join("").trim();
  if (tail) segs.push(tail);
  return segs;
}
/**
 * 修复引号被切到下一段的问题：如果一段以右引号开头，就合并到上一段末尾
 */
function attachClosingQuotes(segs: string[]): string[] {
  const fixed: string[] = [];
  const CLOSING_QUOTES = new Set(['"', "'", "」", "』", "》", "）", "】", "\u201d", "\u2019", "]"]);
  for (const s of segs) {
    if (s && CLOSING_QUOTES.has(s[0]) && fixed.length) {
      fixed[fixed.length - 1] = `${fixed[fixed.length - 1]}${s}`.trim();
    } else {
      fixed.push(s.trim());
    }
  }
  return fixed;
}
/**
 * 合并 <5 字符的超短段到下一段（避免字幕一闪而过）
 */
function mergeShort(segs: string[]): string[] {
  const merged: string[] = [];
  let i = 0;
  while (i < segs.length) {
    const cur = segs[i];
    if (cur.trim().length < 5 && i + 1 < segs.length) {
      segs[i + 1] = `${cur}${segs[i + 1]}`.trim();
      i++;
      continue;
    }
    merged.push(cur);
    i++;
  }
  return merged;
}
function stripTrailingPunct(segs: string[]): string[] {
  return segs
    .map((s) => {
      const t = s.trim();
      if (!t) return "";
      if (t.endsWith("，") || t.endsWith(",") || t.endsWith("。")) return t.slice(0, -1);
      return t.replace(/\s+/g, " ").trim();
    })
    .filter(Boolean);
}

/** 先保留 */
function splitSubtitle(text: string): string[] {
  if (!text.trim()) return [];
  const segs = stripTrailingPunct(mergeShort(attachClosingQuotes(splitProtected(text))));
  return segs.length ? segs : [text.trim()];
}

/**
 * 按 fragment 分配时间的写入逻辑（就是从原 writeSrt 抽出来的）
 * 先保留
 */
function writeSrtFragments(
  lines: string[],
  idx: { value: number },
  start: number,
  end: number,
  fragments: string[],
) {
  const totalDuration = end - start;
  const weights = fragments.map((f) => Math.max(1, f.replace(/\s/g, "").length));
  const totalWeight = weights.reduce((a, b) => a + b, 0);
  let cursor = start,
    allocated = 0;
  for (let f = 0; f < fragments.length; f++) {
    const share =
      f < fragments.length - 1
        ? Math.max(
            200,
            Math.min(
              Math.round((totalDuration * weights[f]) / totalWeight),
              totalDuration - allocated - 100,
            ),
          )
        : Math.max(100, totalDuration - allocated);
    lines.push(
      String(idx.value),
      `${srtTime(cursor)} --> ${srtTime(cursor + share)}`,
      fragments[f],
      "",
    );
    cursor += share;
    allocated += share;
    idx.value++;
  }
}

export function writeSrt(
  segments: (Timing | TranslateSegment)[],
  ctx: TaskCtx,
  outputPath: string,
  useSource?: boolean,
) {
  console.log(`Writing SRT length: ${segments.length}...`);
  const lines: string[] = [];
  let idx = 1;
  for (const item of segments) {
    const start = Math.floor("actual_start" in item ? item?.actual_start : item.start_ms);
    const end = Math.floor("actual_end" in item ? item?.actual_end : item.end_ms);
    // if (end <= start) continue;

    const text = useSource ? (item.text || "").trim() : (item.dst || item.text || "").trim();
    if (!text) continue;
    // 默认一条 SRT（之前由 splitSubtitle 切分的逻辑已抽成 writeSrtFragments）
    lines.push(String(idx), `${srtTime(start)} --> ${srtTime(end)}`, text, "");
    idx++;
  }

  const content = lines.join("\n");
  // 预检：在交给 ffmpeg 前拦截内容问题，避免其报出含糊的 "Unable to open"
  validateSrtContent(content, outputPath);

  writeFile(outputPath, content, ctx);
}

const SRT_TIME_RE = /^\d{2}:\d{2}:\d{2},\d{3} --> \d{2}:\d{2}:\d{2},\d{3}$/;

/**
 * 校验 SRT 文本内容是否可被 ffmpeg/libass 的 subtitles 滤镜正常读取。
 * ffmpeg 在字幕文件内容有问题时往往只报模糊的 "Unable to open"，
 * 这里在写盘前给出明确的错误信息（文件名 + 块号 + 原因）。
 */
export function validateSrtContent(content: string, filePath: string): void {
  const rawLines = content.split("\n");
  const blocks: string[][] = [];
  let current: string[] = [];
  for (const line of rawLines) {
    if (line.trim() === "") {
      if (current.length) blocks.push(current);
      current = [];
    } else {
      current.push(line);
    }
  }
  if (current.length) blocks.push(current);

  if (blocks.length === 0) {
    throw new Error(`SRT 预检失败 (${filePath}): 文件为空，没有任何字幕块`);
  }

  blocks.forEach((block, i) => {
    const blockNo = i + 1;
    if (block.length < 2) {
      throw new Error(
        `SRT 预检失败 (${filePath}): 第 ${blockNo} 块结构非法，期望 "序号 / 时间轴 / 文本"，实际 ${block.length} 行`,
      );
    }
    if (!SRT_TIME_RE.test(block[1])) {
      throw new Error(
        `SRT 预检失败 (${filePath}): 第 ${blockNo} 块时间轴格式非法: "${block[1]}" (应为 HH:MM:SS,mmm --> HH:MM:SS,mmm)`,
      );
    }
    const [s, e] = block[1].split(" --> ").map((t) => {
      const [h, m, rest] = t.split(":");
      const [s2, ms] = rest.split(",");
      return (+h * 3600 + +m * 60 + +s2) * 1000 + +ms;
    });
    if (!(e > s)) {
      throw new Error(
        `SRT 预检失败 (${filePath}): 第 ${blockNo} 块时间轴非法，结束时间必须晚于开始时间 (${block[1]})`,
      );
    }
    // 文本行不得包含 ffmpeg subtitles 滤镜无法处理的 NUL 字符
    for (let li = 2; li < block.length; li++) {
      if (block[li].includes("\u0000")) {
        throw new Error(`SRT 预检失败 (${filePath}): 第 ${blockNo} 块文本含非法 NUL 字符`);
      }
    }
  });
}
